Fixes 942 a Member graduates to Alumni and their subscription ends

ShadyNagy opened the pull request:
## Description

We have implemented important updates to the subscription cancellation process to improve the application's behavior when dealing with alumni users.

## Changes

- A new method `FindUserIsAlumniByEmailAsync` has been added to the `UserLookupService` class. This method identifies if a user is an alumni based on their email.
- The `HandleCustomerSubscriptionCancelledAtPeriodEndAsync` method in the main class has been updated to utilize the `IsAlumniAsync` method. If a user is identified as an alumni, the cancellation process is halted.
- The `HandleCustomerSubscriptionEndedAsync` method in the main class has been updated to utilize the `IsAlumniAsync` method. If a user is identified as an alumni, the end process is halted.
- The `HandleCustomerSubscriptionRenewedAsync` method in the main class has been updated to utilize the `IsAlumniAsync` method. If a user is identified as an alumni, the renewal process is halted.

## Impact

This update refines the subscription cancellation process, especially for alumni users, leading to a smoother and more logical application flow.

1. `HandleCustomerSubscriptionCancelledAtPeriodEndAsync`: With the inclusion of the IsAlumni check, if a customer who cancels their subscription is an alumni, the process will terminate early, and no cancellation email will be sent. Also, no billing activity for the cancellation will be recorded. This is because alumni members have lifetime access even if their subscription is cancelled.

2. `HandleCustomerSubscriptionEndedAsync`: Similarly, if an alumni's subscription ends, they will not be removed from the member role, and no cancellation email will be sent. Also, no member subscription ending billing activity will be recorded, ensuring that the alumni status remains unaffected.

3. `HandleCustomerSubscriptionRenewedAsync`: If an alumni renews their subscription, the function will return early and not perform any actions. This includes not extending their subscription (since it's lifetime), not recording the payment amount, and not adding a billing activity. This is because the renewal process is unnecessary for alumni members as they retain lifetime access.

Overall, these changes ensure that the alumni status of a member is respected, and unnecessary actions such as sending cancellation emails, removing roles, or recording billing activities are not executed for alumni members.

## Checklist

- [x] Implemented `FindUserIsAlumniByEmailAsync` method
- [x] Updated `HandleCustomerSubscriptionCancelledAtPeriodEndAsync` method to consider alumni users
- [x] Code compiles correctly

## Related Issue
closes #942

ardalis commented:
You say "the cancellation process is halted". Just to clarify, their Stripe subscription is still halted (so they are no longer charged) - they just aren't removed from all roles in devBetter, right?

ShadyNagy commented:
@ardalis 
Yes, you are correct, so I changed the impact to be more clear and updated the code:

> ## Impact

> This update refines the subscription cancellation process, especially for alumni users, leading to a smoother and more logical application flow.

> 1. `HandleCustomerSubscriptionCancelledAtPeriodEndAsync`: With the inclusion of the IsAlumni check, if a customer who cancels their subscription is an alumni, the process will terminate early, and no cancellation email will be sent. Also, no billing activity for the cancellation will be recorded. This is because alumni members have lifetime access even if their subscription is cancelled.

> 2. `HandleCustomerSubscriptionEndedAsync`: Similarly, if an alumni's subscription ends, they will not be removed from the member role, and no cancellation email will be sent. Also, no member subscription ending billing activity will be recorded, ensuring that the alumni status remains unaffected.

> 3. `HandleCustomerSubscriptionRenewedAsync`: If an alumni renews their subscription, the function will return early and not perform any actions. This includes not extending their subscription (since it's lifetime), not recording the payment amount, and not adding a billing activity. This is because the renewal process is unnecessary for alumni members as they retain lifetime access.

> Overall, these changes ensure that the alumni status of a member is respected, and unnecessary actions such as sending cancellation emails, removing roles, or recording billing activities are not executed for alumni members.
